File: leetcode/Intersection_of_Two_Linked_Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

/**
 * This problem is similar to the LCA problem with parent link.
 *
 * Similar LCA problems:
 * - find the LCA of two given nodes in a BST, then we can perform pre-order traversal
 *   to find the FIRST node which is between [a, b]
 *
 *      * if x > a && x > b, then x = x->left
 *      * if x < a && x < b, then x = x->right
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        size_t lenA = list_length(headA);
        size_t lenB = list_length(headB);
        
        ListNode *p = headA;
        ListNode *q = headB;
        if(lenA < lenB){
            q = advance_steps(q, lenB-lenA);
        }else{
            p = advance_steps(p, lenA-lenB);
        }
        
        while(p && q){
            if(p == q) break;
            else{
                p = p->next;
                q = q->next;
            }
        }
        
        if(p) return p;
        else return nullptr;
    }
    
private:
    ListNode* advance_steps(ListNode *head, size_t steps){
        if(!head || steps <= 0) return head;
        
        while(head && steps > 0){
            head = head->next;
            steps--;
        }
        
        return head;
    }

    size_t list_length(ListNode *head){
        if(!head) return 0;
        
        ListNode *p = head;
        size_t len = 0;
        while(p){
            len++;
            p = p->next;
        }
        return len;
    }
};

File: leetcode/java/IntersectionofTwoLinkedLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode ha, ListNode hb) {
        if(ha == null || hb == null) return null;
        
        ListNode p = ha;
        ListNode q = hb;
        while(p != null && q != null) {
            // intersection during traversal
            if(p == q) return p;
            p = p.next;
            q = q.next;
        }
        if(p == null && q == null) return null;
        if(p == null) p = hb;
        else q = ha;
        
        while(p != null && q != null) {
            p = p.next;
            q = q.next;
        }
        
        if(p == null) p = hb;
        if(q == null) q = ha;
        
        while(p != null && q != null) {
            if (p == q) return p;
            p = p.next;
            q = q.next;
        }
        
        return null;
        
    }
}
